pkg/speaker/bgp/table/table.go

```diff
@@ -0,0 +1,339 @@
+package table
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/openelb/openelb/pkg/speaker/bgp/config"
+	api "github.com/osrg/gobgp/api"
+	"github.com/osrg/gobgp/pkg/packet/bgp"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	GLOBAL_RIB_NAME = "global"
+)
+
+type DefinedType int
+
+const (
+	DEFINED_TYPE_PREFIX DefinedType = iota
+	DEFINED_TYPE_NEIGHBOR
+	DEFINED_TYPE_TAG
+	DEFINED_TYPE_AS_PATH
+	DEFINED_TYPE_COMMUNITY
+	DEFINED_TYPE_EXT_COMMUNITY
+	DEFINED_TYPE_LARGE_COMMUNITY
+	DEFINED_TYPE_NEXT_HOP
+)
+
+type RouteType int
+
+const (
+	ROUTE_TYPE_NONE RouteType = iota
+	ROUTE_TYPE_ACCEPT
+	ROUTE_TYPE_REJECT
+)
+
+type PolicyDirection int
+
+const (
+	POLICY_DIRECTION_NONE PolicyDirection = iota
+	POLICY_DIRECTION_IMPORT
+	POLICY_DIRECTION_EXPORT
+)
+
+type MatchOption int
+
+const (
+	MATCH_OPTION_ANY MatchOption = iota
+	MATCH_OPTION_ALL
+	MATCH_OPTION_INVERT
+)
+
+func (o MatchOption) ConvertToMatchSetOptionsRestrictedType() config.MatchSetOptionsRestrictedType {
+	switch o {
+	case MATCH_OPTION_ANY:
+		return config.MATCH_SET_OPTIONS_RESTRICTED_TYPE_ANY
+	case MATCH_OPTION_INVERT:
+		return config.MATCH_SET_OPTIONS_RESTRICTED_TYPE_INVERT
+	}
+	return "unknown"
+}
+
+func NewMatchOption(c interface{}) (MatchOption, error) {
+	switch t := c.(type) {
+	case config.MatchSetOptionsType:
+		t = t.DefaultAsNeeded()
+		switch t {
+		case config.MATCH_SET_OPTIONS_TYPE_ANY:
+			return MATCH_OPTION_ANY, nil
+		case config.MATCH_SET_OPTIONS_TYPE_ALL:
+			return MATCH_OPTION_ALL, nil
+		case config.MATCH_SET_OPTIONS_TYPE_INVERT:
+			return MATCH_OPTION_INVERT, nil
+		}
+	case config.MatchSetOptionsRestrictedType:
+		t = t.DefaultAsNeeded()
+		switch t {
+		case config.MATCH_SET_OPTIONS_RESTRICTED_TYPE_ANY:
+			return MATCH_OPTION_ANY, nil
+		case config.MATCH_SET_OPTIONS_RESTRICTED_TYPE_INVERT:
+			return MATCH_OPTION_INVERT, nil
+		}
+	}
+	return MATCH_OPTION_ANY, fmt.Errorf("invalid argument to create match option: %v", c)
+}
+
+type Policy struct {
+	Name string
+}
+
+func (p *Policy) ToConfig() *config.PolicyDefinition {
+	return &config.PolicyDefinition{
+		Name: p.Name,
+	}
+}
+
+type PolicyAssignment struct {
+	Name     string
+	Type     PolicyDirection
+	Policies []*Policy
+	Default  RouteType
+}
+
+var _regexpMedActionType = regexp.MustCompile(`([+-]?)(\d+)`)
+
+func toStatementApi(s *config.Statement) *api.Statement {
+	cs := &api.Conditions{}
```

Comment:
Why isn't it like this?
``` golang
marshal, _ := json.Marshal(s) 
stmt  := &api.Statement{}
_ = json.Unmarshal(marshal, stmt)

```

Reply:
GoBGP configuration types are different from gRPC/CRD types. The intermediate conversion must exist.